adapters/beachfront/beachfront.go

```diff
@@ -111,177 +113,208 @@ type BeachfrontResponseSlot struct {
+
+	if len(beachfrontRequests.Banner.Slots) > 0 {
+		bytes, err := json.Marshal(beachfrontRequests.Banner)
+
+		if err == nil {
+			reqs = append(reqs, &adapters.RequestData{
+				Method:  "POST",
+				Uri:     BannerEndpoint,
+				Body:    bytes,
+				Headers: headers,
+			})
+		}
+	}
+
+	return reqs, errs
 }
 
-func (a *BeachfrontAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
-	var beachfrontRequests BeachfrontRequests
-	var reqJSON []byte
-	var uri string
-	var errs = make([]error, 0, len(request.Imp))
-	var err error
-	var imps int
+func preprocess(request *openrtb.BidRequest) (beachfrontReqs beachfrontRequests, errs []error) {
 
-	uri = getEndpoint(request)
+	var videoImps = make([]openrtb.Imp, 0)
+	var bannerImps = make([]openrtb.Imp, 0)
 
-	beachfrontRequests, errs, imps = preprocess(request, uri)
+	weGotNothing := true
 
-	if uri == VideoEndpoint {
-		reqJSON, err = json.Marshal(beachfrontRequests.Video)
-		uri = uri + beachfrontRequests.Video.AppId + VideoEndpointSuffix
-	} else {
-		/*
-			We will get here if request contains no Video imps, though it might have
-			Audio or Native imps as well as banner.
-		*/
-		reqJSON, err = json.Marshal(beachfrontRequests.Banner)
+	for i := 0; i < len(request.Imp); i++ {
+		if request.Imp[i].Banner != nil {
+			weGotNothing = false
+			bannerImps = append(bannerImps, request.Imp[i])
+		}
+
+		if request.Imp[i].Video != nil {
+			weGotNothing = false
+			videoImps = append(videoImps, request.Imp[i])
+		}
 	}
 
-	if imps == 0 {
-		errs = append(errs, errors.New("no valid impressions were found"))
-		return nil, errs
+	if weGotNothing {
+		errs = append(errs, errors.New("no valid impressions were found in the request"))
+		return
 	}
 
-	if err != nil {
-		errs = append(errs, err)
-		return nil, errs
+	request.Imp = make([]openrtb.Imp, 0)
+
+	if len(bannerImps) > 0 {
+		request.Imp = bannerImps
+		beachfrontReqs.Banner, errs = getBannerRequest(request)
 	}
-	// ------------------------------------
 
-	headers := http.Header{}
-	headers.Add("Content-Type", "application/json;charset=utf-8")
-	headers.Add("Accept", "application/json")
+	if len(videoImps) > 0 {
+		request.Imp = videoImps
+		beachfrontReqs.Video, errs = getVideoRequests(request)
+	}
 
-	if request.Device != nil {
-		addHeaderIfNonEmpty(headers, "User-Agent", request.Device.UA)
-		addHeaderIfNonEmpty(headers, "X-Forwarded-For", request.Device.IP)
-		addHeaderIfNonEmpty(headers, "Accept-Language", request.Device.Language)
-		if request.Device.DNT != nil {
-			addHeaderIfNonEmpty(headers, "DNT", strconv.Itoa(int(*request.Device.DNT)))
-		}
+	return
+}
+
+func newBeachfrontBannerRequest() beachfrontBannerRequest {
+	r := beachfrontBannerRequest{}
+	r.AdapterName = beachfrontAdapterName
+	r.AdapterVersion = beachfrontAdapterVersion
+
+	return r
+}
+
+func newBeachfrontVideoRequest() beachfrontVideoRequest {
+	r := beachfrontVideoRequest{}
+	r.Cur = append(r.Cur, "USD")
```

Comment:
Will it be an issue if `r.Cur` already contains a "USD" entry?

Reply:
It would be an issue, but I don't think it's possible, as r is a spanking new beachfrontVideoRequest, but I added a test for it anyway.